tensorboard/manager.py

```diff
@@ -0,0 +1,149 @@
+  if not os.path.exists(path):
+    os.mkdir(path)
+  return path
+
+
+def _info_file_path():
+  """Get path to info file for the current process."""
+  return os.path.join(_get_info_dir(), "pid-%d.info" % os.getpid())
+
+
+def write_info_file(tensorboard_info):
+  """Write TensorboardInfo to the current process's info file.
+
+  This should be called by `main` once the server is ready. When the
+  server shuts down, `remove_info_file` should be called.
+  """
+  payload = json.dumps(tensorboard_info._asdict(), sort_keys=True, indent=4)
+  with open(_info_file_path(), "w") as outfile:
+    outfile.write("%s\n" % payload)
+
+
+def remove_info_file():
+  """Remove the current process's TensorboardInfo file, if it exists.
+
+  If the file does not exist, no action is taken and no error is raised.
+  """
+  try:
+    os.unlink(_info_file_path())
+  except OSError as e:
+    if e.errno == errno.ENOENT:
+      # The user may have wiped their temporary directory or something.
+      # Not a problem: we're already in the state that we want to be in.
+      pass
+    else:
+      raise
+
+
+def cache_key(working_directory, arguments):
+  """Compute a `TensorboardInfo.cache_key` field.
+
+  Args:
+    working_directory: The directory from which TensorBoard was launched
+      and relative to which paths like `--logdir` and `--db` are
+      resolved.
+    arguments: The command-line args to TensorBoard: `sys.argv[1:]`.
+      Should be a list (or tuple), not an unparsed string. If you have a
+      raw shell command, use `shlex.split` before passing it to this
+      function.
+
+  Returns:
+    A string such that if two (prospective or actual) TensorBoard
+    invocations have the same cache key then it is safe to use one in
+    place of the other.
+  """
+  if not isinstance(arguments, (list, tuple)):
+    raise TypeError(
+        "'arguments' should be a list of arguments, but found: %r "
+        "(use `shlex.split` if given a string)"
+        % (arguments,)
+    )
+  datum = {"working_directory": working_directory, "arguments": arguments}
+  return base64.b64encode(
+      json.dumps(datum, sort_keys=True, separators=(",", ":"))
+  )
+
+
+def get_all():
+  """Read TensorboardInfo values for running TensorBoard processes.
+
+  This function may not provide a perfect snapshot of the set of running
+  processes. Its result set may be incomplete if the user has cleaned
+  their /tmp/ directory while TensorBoard processes are running. It may
+  contain extraneous entries if TensorBoard processes exited uncleanly
+  (e.g., with SIGKILL).
+
+  Returns:
+    A list of `TensorboardInfo` objects.
+  """
+
+  info_dir = _get_info_dir()
+  results = []
+  for filename in os.listdir(info_dir):
+    filepath = os.path.join(info_dir, filename)
+    with open(os.path.join(info_dir, filepath)) as infile:
+      contents = infile.read()
+    try:
+      info = TensorboardInfo(**json.loads(contents))
+    except Exception:
```

Comment:
I haven't reviewed much Python, but is it accepted practice to catch all Exceptions? In Java the custom is to catch specific exceptions.

Reply:
It’s rarely appropriate to use a bare `except:` block, because that
catches `KeyboardInterrupt` and `SystemExit`. An `except Exception:` can
be fine, though here `except ValueError, TypeError:` would probably be
better, yes. I don’t think that it’s _that_ bad given that the body of
the `try` is so short, but it’s true that even a `NameError` and
`AttributeError` are `Exception`s, so restricting has value.

Reply:
IMO it's still best not to `except Exception`.  Catching NameError and friends is just hardly ever what you want to do.

The JSON decoding should only raise ValueError.  For the call to TensorboardInfo, IMO it is a bit better to explicitly extract the fields from the JSON result rather than relying on the namedtuple constructor.  That would also let you do some minimal type-checking like ensuring that the string fields are indeed strings and int fields are indeed ints.